internal/auth/providers/azure_graph.go

```diff
@@ -0,0 +1,170 @@
+	}
+	return &AzureGraphService{
+		client:               client,
+		groupMembershipCache: memberCache,
+		groupNameCache:       nameCache,
+	}
+}
+
+// GetGroups lists groups user belongs to.
+func (gs *AzureGraphService) GetGroups(email string) ([]string, error) {
+	if gs.client == nil {
+		return []string{}, errors.New("oauth client must be configured")
+	}
+	if email == "" {
+		return []string{}, errors.New("missing email")
+	}
+
+	var wg sync.WaitGroup
+	var mtx sync.Mutex
+	var err error
+	groupNames := make([]string, 0)
+	// See: https://developer.microsoft.com/en-us/graph/docs/api-reference/beta/api/user_getmembergroups
+	requestBody := `{"securityEnabledOnly": false}`
+	requestURL := fmt.Sprintf("https://graph.microsoft.com/beta/users/%s/getMemberGroups", url.PathEscape(email))
+	for {
+		groupResponse, err := gs.client.Post(requestURL, "application/json", strings.NewReader(requestBody))
+		if err != nil {
+			return []string{}, err
+		}
+
+		groupData := struct {
+			Next  string   `json:"@odata.nextLink"`
+			Value []string `json:"value"`
+		}{}
+
+		body, err := ioutil.ReadAll(groupResponse.Body)
+		if err != nil {
+			return []string{}, err
+		}
+		if groupResponse.StatusCode >= 400 {
+			return []string{}, fmt.Errorf("api error: %s", string(body))
+		}
+
+		err = json.Unmarshal(body, &groupData)
+		if err != nil {
+			return []string{}, err
+		}
+
+		for _, groupID := range groupData.Value {
+			wg.Add(1)
+			id := groupID
+			go func(wg *sync.WaitGroup) {
+				defer wg.Done()
+
+				var name string
+				// check the cache for the group name first
+				cachedName, ok := gs.groupNameCache.Get(id)
+				if !ok {
+					// didn't have the group name, make concurrent API call to fetch it
+					name, err = gs.getGroupName(id)
+					if err == nil {
+						// got the name ok, populate the cache
+						gs.groupNameCache.Add(id, name)
+					}
+				} else {
+					// cache hit
+					name = cachedName.(string)
+				}
+				mtx.Lock()
+				groupNames = append(groupNames, name)
+				mtx.Unlock()
+			}(&wg)
+		}
+
+		if groupData.Next != "" {
+			requestURL = groupData.Next
+		} else {
+			break
+		}
+	}
+	wg.Wait()
+	if err != nil {
+		return []string{}, err
+	}
+
+	return groupNames, nil
+}
+
+// getGroupName returns the group name, preferentially pulling from cache
+func (gs *AzureGraphService) getGroupName(id string) (string, error) {
+	if gs.client == nil {
+		return "", errors.New("oauth client must be configured")
+	}
+	// See: https://developer.microsoft.com/en-us/graph/docs/api-reference/v1.0/api/group_get
+	requestURL := fmt.Sprintf("https://graph.microsoft.com/v1.0/groups/%s", url.PathEscape(id))
+	groupMetaResponse, err := gs.client.Get(requestURL)
+	if err != nil {
+		return "", err
+	}
+
+	groupMetadata := struct {
+		DisplayName string `json:"displayName"`
+	}{}
+
+	body, err := ioutil.ReadAll(groupMetaResponse.Body)
+	if err != nil {
+		return "", err
+	}
+	if groupMetaResponse.StatusCode >= 400 {
```

Comment:
Also here: The API documentation states that a successful response will return code 200, so lets explicitly compare against `http.StatusOK`.